src/parser.rs

```diff
@@ -0,0 +1,106 @@
+//! This module provides some mild integration between the html5ever parser and async I/O.
+
+use std::io;
+
+use html5ever::driver::{self, Parser};
+use html5ever::tendril::{ByteTendril, TendrilSink};
+use markup5ever_rcdom::{Handle, RcDom};
+use tokio::io::{AsyncRead, AsyncReadExt};
+
+async fn parse_internal_async<R: AsyncRead + Unpin>(
+    parser: Parser<RcDom>,
+    mut r: R,
+) -> io::Result<Handle> {
+    let mut tendril_sink = parser.from_utf8();
+
+    // This draws on the structure of the sync tendril read_from.
+    const BUFFER_SIZE: u32 = 128 * 1024;
+    'read: loop {
+        let mut tendril = ByteTendril::new();
+        unsafe {
+            tendril.push_uninitialized(BUFFER_SIZE);
+        }
+        loop {
+            match r.read(&mut tendril).await {
+                Ok(0) => break 'read,
+                Ok(n) => {
+                    tendril.pop_back(BUFFER_SIZE - n as u32);
+                    tendril_sink.process(tendril);
+                    break;
+                }
+                Err(ref e) if e.kind() == io::ErrorKind::Interrupted => {}
```

Comment:
This is a common pattern with I/O APIs on POSIX platforms. When a sync syscall is interrupted by a signal handler before any data is read from certain file descriptors, it may return EINTR, which signifies that the operation was interrupted and it's left to userspace to reattempt. See, e.g., https://man7.org/linux/man-pages/man2/read.2.html#:~:text=EINTR%20%20The%20call%20was%20interrupted%20by%20a%20signal%20before%20any%20data%20was%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20read%3B%20see%20signal(7).

This is presumably just a reflection of that API design choice into the Rust std::io module.

Reply:
Ohh yeah, no I've definitely seen this before in Chromium as well: https://source.chromium.org/chromium/chromium/src/+/main:base/posix/eintr_wrapper.h;l=28;drc=e4622aaeccea84652488d1822c28c78b7115684f. I should've caught that :)